assign-facets.lua

```diff
@@ -0,0 +1,217 @@
+-- Change the facets of a unit.
+--@ module = true
+
+local help = [====[
+
+assign-facets
+=============
+A script to change the facets (traits) of a unit.
+
+Facets are defined with a token and a number from -3 to 3, which describes
+the different levels of facets strength, as explained here:
+https://dwarffortresswiki.org/index.php/DF2014:Personality_trait#Facets
+
+========  =========
+Strength  Effect
+========  =========
+3         Highest
+2         Very High
+1         High
+0         Neutral
+-1        Low
+-2        Very Low
+-3        Lowest
+========  =========
+
+Resetting a facet means setting it to a level that does not
+trigger a report in the "Thoughts and preferences" screen.
+
+Usage:
+
+:``-help``:
+                    print the help page.
+
+:``-unit <UNIT_ID>``:
+                    set the target unit ID. If not present, the
+                    currently selected unit will be the target.
+
+:``-beliefs <FACET LEVEL [FACET LEVEL] [...]>``:
+                    the facets to modify and their levels. The
+                    valid facet tokens can be found in the wiki page
+                    linked above; level values range from -3 to 3.
+
+:``-reset``:
+                    reset all facets to a neutral level. If the script is
+                    called with both this option and a list of facets/levels,
+                    first all the unit facets will be reset and then those
+                    facets listed after ``-facets`` will be modified.
+
+Example:
+
+``-reset -facets HATE_PROPENSITY -2 CHEER_PROPENSITY -1``
+    Resets all the unit facets, then sets the listed facets to the following values:
+
+    * Hate propensity: a value between 10 and 24 (level -2);
+    * Cheer propensity: a value between 25 and 39 (level -1).
+
+    The final result (for a dwarf) will be:
+    ``She very rarely develops negative feelings
+    toward things. She is rarely happy or enthusiastic,
+    and she is conflicted by this as she values
+    parties and merrymaking in the abstract.``
+
+    Note that the facets are compared to the beliefs,
+    and if conflicts arise they will be reported.
+]====]
+
+local valid_args = {
```

Comment:
There's a `utils.processArgs()` function that does something pretty similar to this - have you tried that?

Reply:
Yes, I thought about adding some of these scripts to gui/gm-unit: I will probably do it in the future, thanks for the suggestion.

And yes, I have tried the `utils.processArgs()` function but it took me a while to understand how to pass lists of different parameters to it: I had to analyse the source code to be sure and so I thought that maybe I should implement it in a different way. Anyway, I will be more than happy to change the arguments parsing function of my scripts to call the `utils` standard function in order to standardise the code between different scripts in this repository, if you prefer so.

Reply:
I just noticed there was a decent amount of argument-parsing code that you'd duplicated between scripts, and was thinking that could clean it up. It's possible that yours behaves differently from processArgs, though, but it looks pretty similar.

Reply:
Ok, I cleaned the argument-parsing code a little bit and performed a few tests: everything seems to be working nicely.
At the moment, I am not so confident about the `assign-profile.lua` script: it feels like it needs more tests, but the other scripts are independent from one another and can be used.